app/services/shop_products_service.rb

```diff
@@ -0,0 +1,76 @@
+# Returns a (paginatable) AR object for the products or variants in stock for a given shop and OC.
+# The stock-checking includes on_demand and stock level overrides from variant_overrides.
+
+class ShopProductsService
+  def initialize(distributor, order_cycle)
+    @distributor = distributor
+    @order_cycle = order_cycle
+  end
+
+  def products_relation
+    Spree::Product.where(id: stocked_products)
+  end
+
+  def variants_relation
+    @order_cycle.
+      variants_distributed_by(@distributor).
+      merge(stocked_variants_and_overrides)
+  end
+
+  private
+
+  def stocked_products
+    @order_cycle.
+      variants_distributed_by(@distributor).
+      merge(stocked_variants_and_overrides).
+      select("DISTINCT spree_variants.product_id")
+  end
+
+  def stocked_variants_and_overrides
+    Spree::Variant.
+      joins("LEFT OUTER JOIN variant_overrides ON variant_overrides.variant_id = spree_variants.id
+            AND variant_overrides.hub_id = #{@distributor.id}").
+      joins(:stock_items).
+      where(query_stock_with_overrides)
+  end
+
+  def query_stock_with_overrides
+    "( #{variant_not_overriden} AND ( #{variant_on_demand} OR #{variant_in_stock} ) )
+      OR ( #{variant_overriden} AND ( #{override_on_demand} OR #{override_in_stock} ) )
+      OR ( #{variant_overriden} AND ( #{override_on_demand_null} AND #{variant_on_demand} ) )
+      OR ( #{variant_overriden} AND ( #{override_on_demand_null}
+                                      AND #{variant_not_on_demand} AND #{variant_in_stock} ) )"
+  end
```

Comment:
we'll need to take a close it to this particular bit of the query. Definitely faster than N+1 though.